test/service/ratelimit_test.go

```diff
@@ -176,6 +187,99 @@ func TestService(test *testing.T) {
 	t.assert.EqualValues(1, t.statStore.NewCounter("config_load_error").Value())
 }
 
+func TestServiceWithCustomHeaders(test *testing.T) {
+	os.Setenv("LIMIT_LIMIT_HEADER", "A-Ratelimit-Limit")
+	os.Setenv("LIMIT_REMAINING_HEADER", "A-Ratelimit-Remaining")
+	os.Setenv("LIMIT_RESET_HEADER", "A-Ratelimit-Reset")
+	defer func() {
+		os.Unsetenv("LIMIT_LIMIT_HEADER")
+		os.Unsetenv("LIMIT_REMAINING_HEADER")
+		os.Unsetenv("LIMIT_RESET_HEADER")
+	}()
+
+	t := commonSetup(test)
+	defer t.controller.Finish()
+	service := t.setupBasicService()
+
+	// Config reload.
+	barrier := newBarrier()
+	t.configLoader.EXPECT().Load(
+		[]config.RateLimitConfigToLoad{{"config.basic_config", "fake_yaml"}}, gomock.Any()).Do(
+		func([]config.RateLimitConfigToLoad, stats.Manager) { barrier.signal() }).Return(t.config)
+	t.runtimeUpdateCallback <- 1
+	barrier.wait()
+
+	// Make request
+	request := common.NewRateLimitRequest(
+		"different-domain", [][][2]string{{{"foo", "bar"}}, {{"hello", "world"}}}, 1)
+	limits := []*config.RateLimit{
+		config.NewRateLimit(10, pb.RateLimitResponse_RateLimit_MINUTE, t.statsManager.NewStats("key"), false),
+		nil}
+	t.config.EXPECT().GetLimit(nil, "different-domain", request.Descriptors[0]).Return(limits[0])
+	t.config.EXPECT().GetLimit(nil, "different-domain", request.Descriptors[1]).Return(limits[1])
+	t.cache.EXPECT().DoLimit(nil, request, limits).Return(
+		[]*pb.RateLimitResponse_DescriptorStatus{{Code: pb.RateLimitResponse_OVER_LIMIT, CurrentLimit: limits[0].Limit, LimitRemaining: 0},
+			{Code: pb.RateLimitResponse_OK, CurrentLimit: nil, LimitRemaining: 0}})
+
+	response, err := service.ShouldRateLimit(nil, request)
+	common.AssertProtoEqual(
+		t.assert,
+		&pb.RateLimitResponse{
+			OverallCode: pb.RateLimitResponse_OVER_LIMIT,
+			Statuses: []*pb.RateLimitResponse_DescriptorStatus{
+				{Code: pb.RateLimitResponse_OVER_LIMIT, CurrentLimit: limits[0].Limit, LimitRemaining: 0},
+				{Code: pb.RateLimitResponse_OK, CurrentLimit: nil, LimitRemaining: 0},
+			},
+			ResponseHeadersToAdd: []*core.HeaderValue{
+				{Key: "A-Ratelimit-Limit", Value: "10"},
+				{Key: "A-Ratelimit-Remaining", Value: "0"},
+				{Key: "A-Ratelimit-Reset", Value: "58"},
```

Comment:
I'm looking at the proto definitions and shouldn't the values like `{Key: "A-Ratelimit-Remaining", Value: "0"}` be set under `LimitRemaining` in pb.RateLimitResponse_DescriptorStatus ?

Reply:
But they are headers so it needs to go into the ResponseHeadersToAdd element